--- a/drivers/net/dsa/mt7530.c
+++ b/drivers/net/dsa/mt7530.c
@@ -2912,6 +2912,7 @@
 /* MMIO (built-in switch) support: MT7988 and Airoha EN7581 */
 static const struct of_device_id mt7988_of_match[] = {
 	{ .compatible = "airoha,en7581-switch", .data = &mt753x_table[ID_EN7581] },
+	{ .compatible = "airoha,an7583-switch", .data = &mt753x_table[ID_EN7581] },
 	{ .compatible = "mediatek,mt7988-switch", .data = &mt753x_table[ID_MT7988] },
 	{ /* sentinel */ }
 };
