--- a/drivers/Kconfig
+++ b/drivers/Kconfig
@@ -226,6 +226,8 @@
 
 source "drivers/interconnect/Kconfig"
 
 source "drivers/counter/Kconfig"
+
+source "drivers/rknpu/Kconfig"
 
 endmenu
--- a/drivers/Makefile
+++ b/drivers/Makefile
@@ -180,9 +180,10 @@
 obj-$(CONFIG_FPGA)		+= fpga/
 obj-$(CONFIG_FSI)		+= fsi/
 obj-$(CONFIG_TEE)		+= tee/
 obj-$(CONFIG_MULTIPLEXER)	+= mux/
 obj-$(CONFIG_UNISYS_VISORBUS)	+= visorbus/
 obj-$(CONFIG_SIOX)		+= siox/
 obj-$(CONFIG_GNSS)		+= gnss/
 obj-$(CONFIG_INTERCONNECT)	+= interconnect/
 obj-$(CONFIG_COUNTER)		+= counter/
+obj-$(CONFIG_ROCKCHIP_RKNPU)	+= rknpu/
